deploy/containers/credhub_setup/auth.go

```diff
@@ -0,0 +1,120 @@
+package main
+
+// This file contains code to deal with UAA / authentication.
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"time"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+func makeHTTPClientWithCA(serverName string, caCert []byte) (*http.Client, error) {
+	certPool, err := x509.SystemCertPool()
+	if err != nil {
+		return nil, fmt.Errorf("could not get system cert pool: %w", err)
+	}
+	ok := certPool.AppendCertsFromPEM(caCert)
+	if !ok {
+		fmt.Printf("Warning: failed to add CA certs for %s\n", serverName)
+	}
+	return &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs:    certPool,
+				ServerName: serverName,
+			},
+		},
+		Timeout: 60 * time.Second,
+	}, nil
+}
+
+// authenticate with UAA, returning the access token and refresh token.
+func authenticate(ctx context.Context, clientID, clientSecret string) (*http.Client, error) {
+	link, err := getCCLinkData(ctx)
+	if err != nil {
+		return nil, err
+	}
+	ccClient, err := makeHTTPClientWithCA(link.CC.InternalServiceHostname, []byte(link.CC.PublicTLS.CACert))
+	if err != nil {
+		return nil, fmt.Errorf("could not make CC HTTP client: %w", err)
+	}
+
+	tokenURL, err := getTokenURL(ctx, link, ccClient)
+	if err != nil {
+		return nil, fmt.Errorf("could not get UAA token URL: %w", err)
+	}
+
+	credentialsConfig := clientcredentials.Config{
+		ClientID:     clientID,
+		ClientSecret: clientSecret,
+		TokenURL:     tokenURL.String(),
+		Scopes:       []string{"cloud_controller.admin"},
+	}
+
+	fmt.Printf("Got UAA token URL: %s\n", tokenURL.String())
+	certPath := filepath.Join(getMountRootFromContext(ctx), "run/uaa-ca-cert/ca.crt")
+	uaaCABytes, err := ioutil.ReadFile(certPath)
```

Comment:
`ioutil.ReadFile` is a hidden dependency.

Reply:
This is getting to the point where it seems like _every_ function call is being considered a hidden dependency? Why isn't `append()` (it allocates memory!) one? I think I may be not interpreting it correctly: is "hidden dependency" some specific jargon with non-obvious meaning?

Reply:
`append` is not a concerning hidden dependency because we don't want to recover if it fails. Failing `append` most likely means we are about to crash the process.

If a function could genuinely fail and we need to take an action on that failure, and that function does not come from either the constructor or the parameter list, it's a hidden dependency that we should care about.